File: CrossCompileApp/composeApp/src/commonMain/kotlin/com/androbrain/crosscompileapp/ui/dashboard/DashboardViewModel.kt
package com.androbrain.crosscompileapp.ui.dashboard

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.androbrain.crosscompileapp.data.repository.BudgetModel
import com.androbrain.crosscompileapp.data.repository.BudgetRepository
import com.androbrain.crosscompileapp.ui.toAmount
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

class DashboardViewModel(
    private val budgetRepository: BudgetRepository,
) : ViewModel() {
    private val _state = MutableStateFlow(DashboardState())
    val state = _state.asStateFlow()

    init {
        observeBudgets()
        observeBalance()
    }

    fun deleteBudget(budget: BudgetModel) {
        viewModelScope.launch { budgetRepository.delete(budget) }
    }

    private fun observeBudgets() {
        budgetRepository.getAll().onEach { budgets ->
            _state.update { state -> state.copy(budgets = budgets.reversed()) }
        }.launchIn(viewModelScope)
    }

    private fun observeBalance() {
        budgetRepository.getAll().onEach { budgets ->
            val balance = budgets.sumOf { budget ->
                if (budget.isSpending) {
                    -1
                } else {
                    1
                } * budget.amount
            }
            _state.update { state -> state.copy(balance = balance.toAmount()) }
        }.launchIn(viewModelScope)
    }
}

File: iOSApp/iOSApp/ui/dashboard/DashboardViewModel.swift
//
//  DashboardViewModel.swift
//  iOSApp
//
//  Created by Michal Ankiersztajn on 14/05/2025.
//

import Foundation

class DashboardViewModel : ObservableObject {
    
    let repository: BudgetRepository
    @Published var budgets: [BudgetModel] = []
    @Published var selectedBudget: BudgetModel? = nil
    @Published var balance: Int64 = 0
    
    init(repository: BudgetRepository) {
        self.repository = repository
        loadBudgets()
    }
    
    func createEmptyBudget() {
        let budget = BudgetModel(amount: 0, content: "", id: UUID(), isSpending: false)
        repository.addOrReplace(budget: budget)
        loadBudgets()
        selectedBudget = budget
    }
    
    func deleteBudget(indexSet: IndexSet) {
        indexSet.forEach { index in
            print(index)
            repository.delete(id: budgets[index].id)
        }
        loadBudgets()
    }
    
    func updateNote(budget: BudgetModel) {
        repository.addOrReplace(budget: budget)
        loadBudgets()
        selectedBudget = nil
    }
    
    private func loadBudgets() {
        budgets = repository.getAll().reversed()
        var balance: Int64 = 0
        budgets.forEach { budget in
            if budget.isSpending {
                balance -= budget.amount
            } else {
                balance += budget.amount
            }
        }
        self.balance = balance
        print(balance)
        print("self" + String(self.balance))
    }
}
